code/modules/atmospherics/machinery/portable/canister.dm

```diff
@@ -482,12 +493,30 @@
 		valve_open = !valve_open
 		timing = FALSE
 
-	// Handle gas transfer.
+	var/turf/location = get_turf(src)
+
+	var/mix_air = FALSE
+	var/pressure = release_pressure
+	var/gas_mix = holding?.air_contents
+	var/air_update = FALSE
+
 	if(valve_open)
-		var/turf/T = get_turf(src)
-		var/datum/gas_mixture/target_air = holding ? holding.air_contents : T.return_air()
+		mix_air = TRUE
+
+	// When at least 10% of integrity is lost it starts checking for leaking
+	if(obj_integrity < max_integrity * 0.9)
+		var/leak_chance = (1 - obj_integrity / max_integrity) * 100
+		if(prob(leak_chance))
+			mix_air = TRUE
+			pressure = air_contents.return_pressure() / 10
+			gas_mix = location.return_air()
+			air_update = TRUE
+
+	// Handle gas transfer.
+	if(mix_air)
+		var/datum/gas_mixture/target_air = gas_mix || location.return_air()
 
-		if(air_contents.release_gas_to(target_air, release_pressure) && !holding)
+		if(air_contents.release_gas_to(target_air, pressure) && (!holding || air_update))
```

Comment:
Sorry if this is obvious, but why does air_update need to exist? I don't quite understand

Reply:
When a canister releases gas normally, it will share that gas with something it's attached to if anything. Air leaks however directly go to the turf's air mix instead, even if something is attached to it. Previously this was only checking if something was attached to the canister and if not only then would the turf's air be updated, but with leaks the turf air potentially has to be updated even if something is attached to the canister.

Reply:
ah I see, thank you